### ISSUE TITLE
Filter secrets from build logs

### ISSUE THREAD

**Issue by azz** [buildkite/agent#982]
We add some secrets to our builds using the `environment` hook.

Currently, a pipeline containing

```yaml
steps:
  - command: env
```

will list out all of the secrets in the build log. Not great!

We're going to look at maybe using the `command` hook to replace the secrets with `******` to prevent leaks, but thought it was worth raising an issue here to see if it's a more general issue that could be solved?

**Comment by GreyKn** [comment]
Why run `env` in the first place as a command?

If you need to read values from it for some reason, you could just pipe into grep with -v to skip secrets

**Comment by azz** [comment]
It's not that `env` is used as a command, it can be used to exfiltrate secrets from the environment, either intentionally or accidentally. 

One use case would be turning on `set -x` in bash, a command that looked like this:

```
set -x
curl "https://${FOO_USER}:${FOO_PASSWORD}/"
```

would leak the secrets to the build log.

**Comment by lox** [comment]
Filtering secrets from logs is unfortunately very difficult to do comprehensively, especially if your aim is to prevent exfiltration. Attackers can generally work around any mechanism you have in place, and in some cases it can make it easier to enumerate secrets for them. 

Filtering out accidentally disclosed secrets could be something we could look at, either by filtering out values that are exported from an `environment` hook, or perhaps detecting high-entropy strings. If we get a chance we'll prototype something up, otherwise we'd love a PR!

**Comment by azz** [comment]
What about a simple `BUILDKITE_SECRETS` variable that gets sets during the `environment` hook to a list of environment variables whose values should be considered "secrets".

e.g., `environment`:

```sh
GITHUB_ACCESS_TOKEN=$(get_secret 'github-access-token')
export GITHUB_ACCESS_TOKEN

export BUILDKITE_SECRETS='GITHUB_ACCESS_TOKEN'
```

This way the agent can filter out any secret values from `stdout`/`stderr`. 

> in some cases it can make it easier to enumerate secrets for them.

This approach _would_ do that, but they're already in the environment, so if an agent is compromised you can just serialise the whole `env` and post it off to some URL anyway.

**Comment by azz** [comment]
Fixed in #1109
